### PR TITLE
Rework odoc rules

### PR DESCRIPTION
**Pull request by lubegasimon**
This PR adds `odoc link` as part of [the documentation generation phases](https://ocaml.github.io/odoc/driver.html#document-generation-phases) and stops `dune` from generating documentation for hidden modules (those with double underscores in their names).

### CODE REVIEW

**Comment by lubegasimon** [comment]
Ping: Anything blocking this PR?

**Comment by jonludlam** [comment]
I think you need to satisfy at least the DCO bot, and there's a conflict. Additionally, I had a go at removing `odocl_alias` and it all seems to work fine: https://github.com/jonludlam/dune/commit/257f58d178caa775179e23700d9362daec31e9ec - I think it's not strictly necessary so unless there's a compelling reason why we need the alias it seems better to remove it. Having said that I'm not sure on whether dune ordinarily has helpful aliases around - perhaps @rgrinberg might have an opinion?

**Comment by lubegasimon** [comment]
> jonludlam@257f58d
I will make fixes and squash commits, thanks!

**Comment by lubegasimon** [comment]
@rgrinberg any blockers to get this merged (except the conflicts)?

**Comment by rgrinberg** [comment]
There are no blockers yet but I haven't reviewed your latest changes. I'll do so today.

**Comment by rgrinberg** [comment]
This PR is still special casing modules with `__` in their name. This behavior is not desirable for the reasons I've outlined above.

**Comment by lubegasimon** [comment]
> This PR is still special casing modules with `__` in their name. This behavior is not desirable for the reasons I've outlined above.

I interpret this as asking authors it concerns to rewrite their dune files, for which I’m not sure how many are willing to do so urgently in favour of this PR.

/c.c @aantron

**Comment by jonludlam** [comment]
I believe right now there won't be any problems, because odoc has some compatibility code in to write an empty file when asked to generate an HTML file for a hidden module. When we remove this I suspect there will be problems because dune will be expecting an output file that won't get written.

**Comment by lubegasimon** [comment]
PR ready for another review!

**Comment by jonludlam** [comment]
Looking good! I still think we should remove the odocl_alias as I suggested above though.

**Comment by lubegasimon** [comment]
> Looking good! I still think we should remove the odocl_alias as I suggested above though.



> Looking good! I still think we should remove the odocl_alias as I suggested above though.

Done! I think @jonludlam will need to sign off the commit I cherry-picked to pass the failing check.

**Comment by lubegasimon** [comment]
It seems the only blocker now is the failing macos-workflow
/c.c @rgrinberg

**Comment by jonludlam** [comment]
Agreed, aside from that this is good to merge as far as I can tell.

File: test/blackbox-tests/test-cases/odoc/odoc-wrapped-lib.t/run.t [hunk 1]
```diff
@@ -0,0 +1,23 @@
+This test generates documentation for non-hidden modules only for a library:
+
+  $ ocamlc -c -bin-annot foo.ml
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/odoc/odoc-wrapped-lib.t/run.t:3]: Why do we need this `ocamlc` call?

Comment (lubegasimon) [reply test/blackbox-tests/test-cases/odoc/odoc-wrapped-lib.t/run.t:3]: Surely, that’s a redundant call! Good catch!

**Pull request merged**